File: 611-valid-triangle-number/611-valid-triangle-number.java
class Solution {
    public int triangleNumber(int[] nums) {
        Arrays.sort(nums);
        int ans=0;
        for(int i=0;i<nums.length;i++)
            for(int j=i+1;j<nums.length;j++){
                int sum=nums[i]+nums[j];
                int l=j+1,r=nums.length-1;
                while(l<=r){
                    int m=(l+r)/2;
                    if(nums[m]>=sum){
                        r=m-1;
                    }else l=m+1;
                }
                ans+=(l-j-1);
            }
        
        return ans;
    }
}

File: 611-valid-triangle-number/611-valid-triangle-number.cpp


class Solution {

public:
    int triangleNumber(vector<int>& nums) {
        sort(nums.begin(),nums.end());

        int ans=0;
        for(int k=nums.size()-1;k>=2;k--){
            int s3 = nums[k];
            int i=0;
            int j=k-1;
            while(i<j){
                if(nums[i]+nums[j]<=s3){
                    i++;
                }
                else{
                    ans+=(j-i);
                    j--;
                }
            }
        }
        return ans;
    }
};

